### ISSUE TITLE
Interoperability of C++ and Swift without Bridging layer Objective-C++

### ISSUE THREAD

**Issue by tigrim** [osmandapp/OsmAnd-iOS#3681]
C++ interoperability is a new feature in Swift 5.9. A great variety of C++ APIs can be called directly from Swift, and select Swift APIs can be used from C++.

TODO:
Check the possibility of integrating a new API and its potential usability.

<img width="688" alt="image" src="https://github.com/osmandapp/OsmAnd-iOS/assets/17082167/773f00bd-6d86-4a1d-beb3-612e407c4438">

https://www.swift.org/documentation/cxx-interop/
https://github.com/apple/swift/blob/main/docs/CppInteroperability/GettingStartedWithC%2B%2BInterop.md

**To do** 
- For 3 days reach as maximum as possible. Document results, explain to other team members how to proceed

**Comment by vshcherb** [comment]
## Research result: Swift/C++ interop is not the right target here (gen by AI)

Measured the actual codebase before spending the 3 days. Conclusion is negative for the stated goal, and the numbers point at a different, larger debt.

### Why direct Swift ↔ C++ interop does not pay off

`core/include` — 286 headers:

- **199** expose Qt types in their public API (`QString`, `QList`, `QHash`);
- **201** expose templates or `std::shared_ptr`;
- protobuf (`google/protobuf/io/coded_stream.h`) is included from public headers;
- the dominant pattern is `PrivateImplementation<T>` pimpl + `Q_DISABLE_COPY_AND_MOVE` (e.g. `Data/ObfReader.h`).

Consequences for Swift 5.9 C++ interop:

1. Qt containers import as opaque non-trivial C++ types. There is no implicit `QString → String` or `QList → Array`. Every call site needs a hand-written conversion shim — i.e. the bridging layer we wanted to remove, relocated into Swift.
2. Class templates need explicit instantiation to be visible to Swift.
3. `PrivateImplementation<T>` types with a deleted copy/move are neither value nor reference types for Swift; exposing them requires `SWIFT_SHARED_REFERENCE` annotations with retain/release functions written **into the core headers**, which are shared with Android.
4. Enabling `-cxx-interoperability-mode` makes ClangImporter parse Qt in C++ mode for the whole importing module, with the corresponding build-time cost.

### The layer we wanted to remove does not actually exist

`Sources/OsmAnd Maps-Bridging-Header.h` exposes **322 Objective-C headers** to Swift, and **none of them transitively includes `OsmAndCore`**. C++ is already kept out of Swift's view by the `*+cpp.h` category pattern (19 files). The isolation membrane is in place and working; there is no Swift code today that needs C++ at all.

### What the numbers actually say

`ios/Sources`, excluding `ThirdPart`:

| | files | LOC |
|---|---|---|
| `.m` | 383 | 72 924 |
| `.mm` | 593 | 270 551 |
| `.mm` containing **no** C++ token at all | **303** | **100 415** |
| files including `OsmAndCore` | 204 | 123 899 |
| files whose class name has a twin in Android (Java/Kotlin) | 295 | 93 136 |

C++ density in the files we think of as "C++ files" is very low:

- `Router/OARoutingHelper.mm` — 9 C++ lines out of 1258
- `Router/OAVoiceRouter.mm` — 10 out of 1037
- `Backup/OABackupHelper.mm` — 1 out of 919
- `Helpers/OAOsmAndFormatter.mm` — 0 out of 931

So the debt is not the Objective-C++ bridge. The debt is **duplication**: ~93k lines of Objective-C that re-implement Android Java classes, three OBF readers (`BinaryMapIndexReader.java`, `core-legacy/native/src/binaryRead.cpp`, `core/src/Data/ObfReader.cpp`) and two routing engines.

### Proposal for this issue

Close #3681 as researched/negative, or reduce it to writing up the above. Do not enable app-wide C++ interop. Track the work below instead.

---

## Work plan

### Step 0 — mechanical, no design decisions (iOS)

- Rename the **303 `.mm` files (100 415 LOC) that contain no C++ token** to `.m`. They currently pay Objective-C++ compile cost for nothing. Measure build time before/after.
- Move `OsmAndCore` includes out of the 80 `.h` files that still leak them, into `*+cpp.h` categories — the pattern already used in 19 places. This is a precondition for everything else.
- Align `CLANG_CXX_LANGUAGE_STANDARD` across targets (currently `gnu++14` on one, `c++17` on another).

### Step 1 — land the shared route-details migration

osmandapp/OsmAnd#25773 and osmandapp/OsmAnd-iOS#5707 establish the pattern to reuse everywhere else: narrow accessor interfaces (`ILocationAccessor`, `IManeuverAccessor`, `IRouteStatisticsAccessor`) plus an eager platform-neutral snapshot, backed by compatibility tests against Android-derived fixtures.

Note the limit of the pattern: eager copying is correct for bounded data such as route details. It must **not** be generalised to OBF reading, which is lazy, unbounded and on the hot path.

### Step 2 — move duplicated logic to `OsmAnd-shared`, in this order

Selection criterion: an Android twin exists **and** C++ density is ~0.

1. **`Backup/` + `Backup/LocalBackup/` + `Backup/Commands/`** — 61 files, 15 559 LOC, near 1:1 with `net.osmand.plus.backup`, 1 C++ line in the whole cluster, nothing of it in `OsmAnd-shared` yet. Highest value: sync/merge divergence between platforms is a live source of bugs.
2. **`QuickAction/Actions/` + `OAMapButtonsHelper`** — 34 files, 4 561 LOC. Many small uniform classes, low risk, good second run to validate the process.
3. **Utility classes that belong in `OsmAnd-java`/shared territory, not in the iOS repo** — `OAApplicationMode` (1080), `OAOsmAndFormatter` (931), `OALocationParser` (594), `SunriseSunset` (518), `OATspAnt` (514), `OALocationConvert` (341), `OACollatorStringMatcher` (260), `QuadTree` (164).
4. **`OsmEdit/OsmData/`** — 8 files, 1 888 LOC, all of it `net.osmand.osm.edit` (`Entity`, `Way`, `Relation`, `OsmBaseStorage`).
5. **`Router/` minus `OARouteProvider.mm`** — ~9 900 of 11 909 LOC. `OARouteProvider` stays: it is where the C++ engine is actually invoked.

Every cluster needs a behaviour diff before the move. Name-based twin detection is a heuristic — the iOS ports have drifted over the years, so compatibility tests in the style of Step 1 are mandatory, not optional.

### Step 3 — decide the OBF reader question (blocks search/POI work)

Options:

- **cinterop over C++ — rejected.** Kotlin/Native cinterop handles C only. A flat `extern "C"` shim over the Qt-based core would be useless on Android (JVM needs JNI), leaving two bindings for one reader.
- **`expect`/`actual` seam — do this now.** Declare a platform-neutral, cursor/visitor-shaped read interface in shared (`searchMapIndex(request, visitor)`, never "return a list", otherwise megabytes get copied at the boundary). `actual` on Android = existing `BinaryMapIndexReader`; on iOS = thin Objective-C wrapper over C++ using the existing `+cpp.h` pattern. Unblocks moving search/POI logic without touching either reader.
- **Port `BinaryMapIndexReader` to common Kotlin — the endgame, gated on a benchmark.** Feasibility is better than it looks: okio 3.9 is already a dependency of `OsmAnd-shared` and its `FileHandle` gives random access on all targets; the reader does manual `readTag()`/`WireFormat` parsing in 28 places, so the vendored patched `com/google/protobuf/CodedInputStream.java` (1011 lines, `RandomAccessFile` + `seek`) can be ported onto okio instead of pulling in a KMP protobuf runtime, and the 81k-line generated `OsmandOdb.java` mostly does not need porting (~44 usages). Scope ≈ 10k lines plus the stream, comparable to the GPX port already in shared (12 461 LOC).

**Spike to run instead of the original 3 days of #3681:** port `CodedInputStream` and the map section of `BinaryMapIndexReader` to Kotlin, then benchmark reading a real OBF against the Java and C++ readers. Kotlin/Native GC behaviour and the absence of value types in the inner loop are the only real risk, and they must be measured before committing. That benchmark is the decision gate for the full port — and for eventually porting the routing engine, which is the largest duplicate of all.

### Step 4 — Swift rewrites, only where there is no Android twin and no C++

- `Purchases/OAProducts.mm` (3227) + `Purchases/OAIAPHelper.mm` (2380) — 5 607 LOC, zero C++, StoreKit. No Android counterpart exists or ever will, so shared Kotlin cannot help; Swift buys StoreKit 2 and async/await. Best return per unit of work in this bucket.
- `Services/OALocationServices.mm` (1049) — CoreLocation, zero C++.
- `Controllers/Cloud/*` — after Step 2.1 only, otherwise it gets rewritten twice.

### Do not touch

- `Controllers/Map/Layers/*` (27 files, 15 594 LOC) — C++ density looks low (118 lines of 2187 in `OAGPXLayer.mm`), but those lines are the entire purpose of the files: symbol providers and `MapRendererView`. Rewriting them in Swift means writing Qt shims by hand, exactly the work this issue set out to avoid.
- `Sources/Core/*` (11 files, 2 680 LOC), `AppHost/OsmAndAppImpl.mm`, `Views/OAMapRendererView.mm` — this is the membrane, and it works.
- `Search/OASearchCoreFactory.mm` (2772) — has an Android twin but depends on OBF reading; blocked on Step 3.
- `Controllers/Panels/OAMapPanelViewController.mm` (4687), `Controllers/Map/OAMapViewController.mm` (4541) — application-wide hubs; only carve pieces off opportunistically during feature work, never rewrite wholesale.
